discovery_engine_core/bindings/src/types/embedding.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! FFI functions for handling embeddings.
+
+use core::document::{Embedding, Embedding1};
+
+use ndarray::Array;
+
+use crate::types::slice::boxed_slice;
+
+/// Creates a rust `Embedding1` with given capacity at given memory address.
+///
+/// The `Embedding1` is created at given address, not it's content.
+///
+/// Returns a pointer to the begin of the buffer of the embedding.
+///
+/// # Safety
+///
+/// - It must be valid to write an `Embedding1` instance to given pointer.
+/// - The passed in slice must represent a `Box<[f32]>` and transfers ownership,
+///   it must be fully initialized.
+#[no_mangle]
+pub unsafe extern "C" fn init_embedding1_at(
+    place: *mut Embedding1,
+    float_slice: *mut f32,
+    len: usize,
+) {
+    let vec = unsafe { boxed_slice::<f32>(float_slice, len).into() };
+    let embedding = Embedding(Array::from_vec(vec));
+    unsafe {
+        place.write(embedding);
+    }
+}
+
+/// Returns a pointer to the begin of the `[f32]` backing the `Embedding1`
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Embedding1`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_embedding1_buffer(place: *mut Embedding1) -> *mut f32 {
+    let embedding = unsafe { &mut *place };
+    embedding.0.as_mut_ptr()
+}
+
+/// Returns len of given embeddings buffer.
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Embedding1`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_embedding1_buffer_len(place: *mut Embedding1) -> usize {
+    //WARNING: This holds for `Embedding1` but not for all possible `ndarray::Array<...>`.
```

Comment:
What this means?

Reply:
For ndarray::Array the array len and arrays buffer len might not be the same (through in this case they are guaranteed to be the same).

The dimension parameter in an `Array` can have a strides which can both skip and overlap the buffer.

Like you can(1) have a buffer of one element but a huge multidimensional array. Just by having a stride of 0 all elements refer to the same element.

(1): Actually I'm not 100% sure if that exact example works with the current `ndarray` interface/library, though it does work with the technology they use and is the reason why some methods returns options where you wouldn't expect it.